File: app/src/main/java/io/stamethyst/backend/RuntimePaths.java
package io.stamethyst.backend;

import android.content.Context;

import java.io.File;

public final class RuntimePaths {
    private RuntimePaths() {
    }

    public static File stsRoot(Context context) {
        return new File(context.getFilesDir(), "sts");
    }

    public static File importedStsJar(Context context) {
        return new File(stsRoot(context), "desktop-1.0.jar");
    }

    public static File importedMtsJar(Context context) {
        return new File(stsRoot(context), "ModTheSpire.jar");
    }

    public static File modsDir(Context context) {
        return new File(stsRoot(context), "mods");
    }

    public static File importedBaseModJar(Context context) {
        return new File(modsDir(context), "BaseMod.jar");
    }

    public static File importedStsLibJar(Context context) {
        return new File(modsDir(context), "StSLib.jar");
    }

    public static File enabledModsConfig(Context context) {
        return new File(stsRoot(context), "enabled_mods.txt");
    }

    public static File preferencesDir(Context context) {
        return new File(stsRoot(context), "preferences");
    }

    public static File betaPreferencesDir(Context context) {
        return new File(stsRoot(context), "betaPreferences");
    }

    public static File mtsGdxApiJar(Context context) {
        return new File(stsRoot(context), "mts-gdx-api.jar");
    }

    public static File mtsStsResourcesJar(Context context) {
        return new File(stsRoot(context), "mts-sts-resources.jar");
    }

    public static File mtsBaseModResourcesJar(Context context) {
        return new File(stsRoot(context), "mts-basemod-resources.jar");
    }

    public static File mtsGdxBridgeJar(Context context) {
        return new File(stsRoot(context), "mts-gdx-bridge.jar");
    }

    public static File mtsLocalJreDir(Context context) {
        return new File(stsRoot(context), "jre");
    }

    public static File mtsLocalJreBinDir(Context context) {
        return new File(mtsLocalJreDir(context), "bin");
    }

    public static File mtsLocalJavaShim(Context context) {
        return new File(mtsLocalJreBinDir(context), "java");
    }

    public static File latestLog(Context context) {
        return new File(stsRoot(context), "latestlog.txt");
    }

    public static File lastCrashReport(Context context) {
        return new File(stsRoot(context), "last_crash_report.txt");
    }

    public static File displayConfigFile(Context context) {
        return new File(stsRoot(context), "info.displayconfig");
    }

    public static File rendererConfigFile(Context context) {
        return new File(stsRoot(context), "renderer_backend.txt");
    }

    public static File componentRoot(Context context) {
        return context.getFilesDir();
    }

    public static File lwjglDir(Context context) {
        return new File(componentRoot(context), "lwjgl3");
    }

    public static File lwjglJar(Context context) {
        return new File(lwjglDir(context), "lwjgl-glfw-classes.jar");
    }

    public static File lwjgl2InjectorDir(Context context) {
        return new File(componentRoot(context), "lwjgl2_methods_injector");
    }

    public static File lwjgl2InjectorJar(Context context) {
        return new File(lwjgl2InjectorDir(context), "lwjgl2_methods_injector.jar");
    }

    public static File bootBridgeDir(Context context) {
        return new File(componentRoot(context), "boot_bridge");
    }

    public static File bootBridgeJar(Context context) {
        return new File(bootBridgeDir(context), "boot-bridge.jar");
    }

    public static File bootBridgeEventsFile(Context context) {
        return new File(stsRoot(context), "boot_bridge_events.log");
    }

    public static File gdxPatchDir(Context context) {
        return new File(componentRoot(context), "gdx_patch");
    }

    public static File gdxPatchJar(Context context) {
        return new File(gdxPatchDir(context), "gdx-patch.jar");
    }

    public static File cacioDir(Context context) {
        return new File(componentRoot(context), "caciocavallo");
    }

    public static File runtimeRoot(Context context) {
        return new File(new File(context.getFilesDir(), "runtimes"), "Internal");
    }

    public static void ensureBaseDirs(Context context) {
        stsRoot(context).mkdirs();
        modsDir(context).mkdirs();
        mtsLocalJreBinDir(context).mkdirs();
        lwjglDir(context).mkdirs();
        lwjgl2InjectorDir(context).mkdirs();
        bootBridgeDir(context).mkdirs();
        gdxPatchDir(context).mkdirs();
        cacioDir(context).mkdirs();
        runtimeRoot(context).mkdirs();
    }
}

File: app/src/main/java/io/stamethyst/backend/core/RuntimePaths.kt
package io.stamethyst.backend.core

import android.content.Context
import java.io.File

object RuntimePaths {
    @JvmStatic
    fun stsRoot(context: Context): File = File(context.filesDir, "sts")

    @JvmStatic
    fun importedStsJar(context: Context): File = File(stsRoot(context), "desktop-1.0.jar")

    @JvmStatic
    fun importedMtsJar(context: Context): File = File(stsRoot(context), "ModTheSpire.jar")

    @JvmStatic
    fun modsDir(context: Context): File = File(stsRoot(context), "mods")

    @JvmStatic
    fun importedBaseModJar(context: Context): File = File(modsDir(context), "BaseMod.jar")

    @JvmStatic
    fun importedStsLibJar(context: Context): File = File(modsDir(context), "StSLib.jar")

    @JvmStatic
    fun enabledModsConfig(context: Context): File = File(stsRoot(context), "enabled_mods.txt")

    @JvmStatic
    fun preferencesDir(context: Context): File = File(stsRoot(context), "preferences")

    @JvmStatic
    fun betaPreferencesDir(context: Context): File = File(stsRoot(context), "betaPreferences")

    @JvmStatic
    fun mtsGdxApiJar(context: Context): File = File(stsRoot(context), "mts-gdx-api.jar")

    @JvmStatic
    fun mtsStsResourcesJar(context: Context): File = File(stsRoot(context), "mts-sts-resources.jar")

    @JvmStatic
    fun mtsBaseModResourcesJar(context: Context): File = File(stsRoot(context), "mts-basemod-resources.jar")

    @JvmStatic
    fun mtsGdxBridgeJar(context: Context): File = File(stsRoot(context), "mts-gdx-bridge.jar")

    @JvmStatic
    fun mtsLocalJreDir(context: Context): File = File(stsRoot(context), "jre")

    @JvmStatic
    fun mtsLocalJreBinDir(context: Context): File = File(mtsLocalJreDir(context), "bin")

    @JvmStatic
    fun mtsLocalJavaShim(context: Context): File = File(mtsLocalJreBinDir(context), "java")

    @JvmStatic
    fun latestLog(context: Context): File = File(stsRoot(context), "latestlog.txt")

    @JvmStatic
    fun lastCrashReport(context: Context): File = File(stsRoot(context), "last_crash_report.txt")

    @JvmStatic
    fun displayConfigFile(context: Context): File = File(stsRoot(context), "info.displayconfig")

    @JvmStatic
    fun rendererConfigFile(context: Context): File = File(stsRoot(context), "renderer_backend.txt")

    @JvmStatic
    fun componentRoot(context: Context): File = context.filesDir

    @JvmStatic
    fun lwjglDir(context: Context): File = File(componentRoot(context), "lwjgl3")

    @JvmStatic
    fun lwjglJar(context: Context): File = File(lwjglDir(context), "lwjgl-glfw-classes.jar")

    @JvmStatic
    fun lwjgl2InjectorDir(context: Context): File = File(componentRoot(context), "lwjgl2_methods_injector")

    @JvmStatic
    fun lwjgl2InjectorJar(context: Context): File =
        File(lwjgl2InjectorDir(context), "lwjgl2_methods_injector.jar")

    @JvmStatic
    fun bootBridgeDir(context: Context): File = File(componentRoot(context), "boot_bridge")

    @JvmStatic
    fun bootBridgeJar(context: Context): File = File(bootBridgeDir(context), "boot-bridge.jar")

    @JvmStatic
    fun bootBridgeEventsFile(context: Context): File = File(stsRoot(context), "boot_bridge_events.log")

    @JvmStatic
    fun gdxPatchDir(context: Context): File = File(componentRoot(context), "gdx_patch")

    @JvmStatic
    fun gdxPatchJar(context: Context): File = File(gdxPatchDir(context), "gdx-patch.jar")

    @JvmStatic
    fun gdxPatchNativesDir(context: Context): File = File(gdxPatchDir(context), "natives")

    @JvmStatic
    fun cacioDir(context: Context): File = File(componentRoot(context), "caciocavallo")

    @JvmStatic
    fun runtimeRoot(context: Context): File = File(File(context.filesDir, "runtimes"), "Internal")

    @JvmStatic
    fun ensureBaseDirs(context: Context) {
        stsRoot(context).mkdirs()
        modsDir(context).mkdirs()
        mtsLocalJreBinDir(context).mkdirs()
        lwjglDir(context).mkdirs()
        lwjgl2InjectorDir(context).mkdirs()
        bootBridgeDir(context).mkdirs()
        gdxPatchDir(context).mkdirs()
        gdxPatchNativesDir(context).mkdirs()
        cacioDir(context).mkdirs()
        runtimeRoot(context).mkdirs()
    }
}
